File: examples/python/dependency-inversion-principle/good.py
import abc

class DBConnectionInterface(metaclass=abc.ABCMeta):
    def connect(self)->str:
        raise NotImplementedError

class MySQLConnection(DBConnectionInterface):
    def connect(self):
        # handle the database connection
        return 'MySQL Database connection'

class MSSQLConnection(DBConnectionInterface):
    def connect(self):
        # handle the database connection
        return 'MSSQL Database connection'

class PasswordReminder:
    def __init__(self, dbConnection: DBConnectionInterface):
        self.dbConnection = dbConnection


'''
Here is an example of a PasswordReminder that connects to a MySQL database:

'''
conn = MySQLConnection()
myPassReminder = PasswordReminder(conn)
print(conn.connect())


conn2 = MSSQLConnection()
msPassReminder = PasswordReminder(conn2)
print(conn2.connect())

File: examples/csharp/DependencyInversionPrinciple/Good.cs
using System;

namespace DIP.Good
{
    public interface IDBConnection
    {
        string Connect();
    }

    public class MySQLConnection: IDBConnection
    {
        private string _name;
        public MySQLConnection() { 
            this._name = "MySQL";
        }

        public string Connect()
        {
            return String.Format("{0} Database connection", this._name);
        }
    }

    public class MSSQLConnection: IDBConnection
    {
        private string _name;
        public MSSQLConnection() {
            this._name = "MSSQL";
         }

        public string Connect()
        {
            return String.Format("{0} Database connection", this._name);
        }
    }


    public class PasswordReminder
    {
        private IDBConnection _connection;
        public PasswordReminder(IDBConnection dbConnection)
        {
            this._connection = dbConnection;
        }
    }
}
